src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -170,6 +170,9 @@ void IEnlistmentNotification.InDoubt(Enlistment enlistment)
 
 			private void ProcessSecondPhase(Enlistment enlistment, bool? success)
 			{
+				// In case of rollback, the prepare phase may have not be run, and we then need to lock as soon as possible.
+				// There is no guarantee the second phase will be run before the completed event, doing that at both places.
+				SafeLockSession();
```

Comment:
This trouble was hidden in test because Dtc test were by-passing the session tracking from `TestBase` by opening session directly from session factory. I have changed that today...